src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs

```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+        {
+            var available = new HashSet<FrameworkRuntimePair>();
+
+            var contentItems = new ContentItemCollection();
+            contentItems.Load(compatibilityData.Files);
+
+
+            if (compatibilityData.TargetLibrary.PackageType.Contains(PackageType.DotnetTool))
+            {
+                foreach (var group in contentItems.FindItemGroups(graph.Conventions.Patterns.ToolsAssemblies))
+                {
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.RuntimeIdentifier, out var ridObj);
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.TargetFrameworkMoniker, out var tfmObj);
+
+                    var tfm = tfmObj as NuGetFramework;
+                    var rid = ridObj as string;
+                    if (tfm?.IsSpecificFramework == true)
+                    {
+                        available.Add(new FrameworkRuntimePair(tfm,rid));
+                    }
+                }
+            }
+
+            return available;
         }
 
+        private async Task VerifyDotnetToolCompatibilityChecks(CompatibilityData compatibilityData, GraphItem<RemoteResolveResult> node, RestoreTargetGraph graph, List<CompatibilityIssue> issues)
+        {
+            var containsDotnetToolPackageType = compatibilityData.TargetLibrary.PackageType.Contains(PackageType.DotnetTool);
+
+            if (compatibilityData.TargetLibrary.PackageType.Count != 1 && containsDotnetToolPackageType)
+            {
+                var issue = CompatibilityIssue.ToolsPackageWithExtraPackageTypes(
+                    new PackageIdentity(node.Key.Name, node.Key.Version));
+
+                issues.Add(issue);
+                await _log.LogAsync(GetErrorMessage(NuGetLogCode.NU1204, issue, graph));
+            }
+
+            if (containsDotnetToolPackageType &&
+                    !(HasCompatibleToolsAssets(compatibilityData.TargetLibrary) || !compatibilityData.Files.Any(p => p.StartsWith("tools/", StringComparison.OrdinalIgnoreCase))))
+            {
+                var available = GetAvailableFrameworkRuntimePairs(compatibilityData, graph);
+                var issue = CompatibilityIssue.IncompatibleToolsPackage(
+                        new PackageIdentity(node.Key.Name, node.Key.Version),
+                        graph.Framework,
+                        graph.RuntimeIdentifier,
+                        available);
+
+                issues.Add(issue);
+                await _log.LogAsync(GetErrorMessage(NuGetLogCode.NU1202, issue, graph));
+            }
+
+            if (ProjectStyle.DotnetToolReference == compatibilityData.PackageSpec.RestoreMetadata?.ProjectStyle)
+            {
+                if (compatibilityData.PackageSpec.GetAllPackageDependencies().Any(e => e.Name.Equals(compatibilityData.TargetLibrary.Name, StringComparison.OrdinalIgnoreCase)))
+                {
+                    if (!containsDotnetToolPackageType)
```

Comment:
change this to :
if(!containsDotnetToolPackageType && compatibilityData.PackageSpec.GetAllPackageDependencies().Any(e => e.Name.Equals(compatibilityData.TargetLibrary.Name, StringComparison.OrdinalIgnoreCase)))

this way it will not have to evaluate the more expensive check if containsDotnetToolPackageType is false.

Reply:
Good catch. Not sure how I missed that simplification.